### PR TITLE
Add data handler for provision requests.

### PR DESCRIPTION
**Pull request by adamsilverstein**
## Summary

Addresses issue #1212

## Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: includes/Modules/Analytics.php [hunk 1]
```diff
@@ -767,6 +775,86 @@ function ( $order_def ) {
...
+
+				$account_ticket = new AccountTicket();
+				$account_ticket->setAccount( $account );
+				$account_ticket->setWebproperty( $property );
+				$account_ticket->setProfile( $profile );
+				$account_ticket->setRedirectUri( $this->get_provisioning_redirect_uri() );
+
+				// Add site id and secret.
+				$creds = $credentials->get();
+				$account_ticket->setSiteId( $creds['oauth2_client_id'] );
```
Comment (adamsilverstein) [inline includes/Modules/Analytics.php:819]: should we check if these are available before using?

File: includes/Modules/Analytics.php [hunk 2]
```diff
@@ -767,6 +775,86 @@ function ( $order_def ) {
...
+					/* translators: %s: Missing parameter name */
+					return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'profileName' ), array( 'status' => 400 ) );
+				}
+				if ( ! isset( $data['timezone'] ) ) {
+					/* translators: %s: Missing parameter name */
+					return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'timezone' ), array( 'status' => 400 ) );
+				}
+				$credentials = new Credentials( new Encrypted_Options( $this->options ) );
+				if ( ! $credentials->has() ) {
+					return new WP_Error( 'missing_credentials', __( 'Missing credentials..', 'google-site-kit' ), array( 'status' => 400 ) );
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:797]: Two periods here :)

File: includes/Modules/Analytics.php [hunk 3]
```diff
@@ -767,6 +775,86 @@ function ( $order_def ) {
...
+				}
+				if ( ! isset( $data['profileName'] ) ) {
+					/* translators: %s: Missing parameter name */
+					return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'profileName' ), array( 'status' => 400 ) );
+				}
+				if ( ! isset( $data['timezone'] ) ) {
+					/* translators: %s: Missing parameter name */
+					return new WP_Error( 'missing_required_param', sprintf( __( 'Request parameter is empty: %s.', 'google-site-kit' ), 'timezone' ), array( 'status' => 400 ) );
+				}
+				$credentials = new Credentials( new Encrypted_Options( $this->options ) );
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:795]: This should use `$this->authentication->credentials()` instead (no need to re-import all these classes and instantiate here).

File: includes/Modules/Analytics.php (hunk 1 above)
Comment (felixarntz) [reply includes/Modules/Analytics.php:819]: No need for that, if they are not there, the previous check `Credentials::has` will have failed.

File: includes/Modules/Analytics.php [hunk 4]
```diff
@@ -481,6 +488,7 @@ protected function get_datapoint_services() {
 			'report'                       => 'analyticsreporting',
 			// POST.
 			'settings'                     => '',
+			'create-account-ticket'        => 'analytics',
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:491]: This needs to use the custom `analyticsprovisioning` instead.

File: includes/Modules/Analytics.php [hunk 5]
```diff
@@ -767,6 +775,86 @@ function ( $order_def ) {
...
+					$account_ticket = $analytics_service->provisioning->createAccountTicket( $account_ticket );
+				} catch ( Exception $e ) {
+					$restore_defer();
+					return $this->exception_to_error( $e, $data->datapoint );
+				}
+				$restore_defer();
+				$account_ticket_id = $account_ticket->getId();
+				if ( empty( $account_ticket_id ) ) {
+					return new WP_Error( 'analytics-provisioning', 'no-account-ticket-id', array( 'status' => 500 ) );
+				}
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:853]: This check can be removed, this should never be empty - if it's empty, an error would be returned instead.

File: includes/Modules/Analytics.php [hunk 6]
```diff
@@ -767,6 +775,86 @@ function ( $order_def ) {
...
+
+				// Cache the create ticket id long enough to verify it upon completion of the terms of service.
+				set_transient( $self::ANALYTICS_PROVISIONING_ACCOUNT_TICKET_ID . '::' . get_current_user_id(), $account_ticket_id, 15 * MINUTE_IN_SECONDS );
+				return $account_ticket;
```
Comment (felixarntz) [inline includes/Modules/Analytics.php:857]: The function in this datapoint needs to be split (we have `create_data_request` and `parse_data_response`).

* `create_data_request` should handle everything up to the request, it needs to `return $analytics_service->provisioning->createAccountTicket( $account_ticket )`. No try-catch and no `$restore_defer` should be in that method, it's all handled by the general datapoint wrapper.
* `parse_data_response` should further handle the `$response`, which will be that account ticket object. So the part where it sets the transient and then just passes through the account ticket should be in the response handler.

See other datapoints that work like that. Essentially every datapoint that as its main purpose makes an API request should be split like that, if it also needs to do something specific with the response. The main purpose of this is to support batching, which we may phase out at some point, but it's still there generally, so let's follow our API here.

File: includes/Modules/Analytics/Account_Ticket.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+use Google\Site_Kit_Dependencies\Google_Service_Analytics_AccountTicket;
+
+/**
+ * Class for the Analytics provisioning Account Ticket object.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class AccountTicket extends Google_Service_Analytics_AccountTicket {
```
Comment (felixarntz) [inline includes/Modules/Analytics/Account_Ticket.php:22]: Just an idea, but maybe call this something like `Proxy_AccountTicket` or similar to further clarify the purpose?

File: includes/Modules/Analytics/Provisioning.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+use Google\Site_Kit\Core\Storage\Setting_With_Legacy_Keys_Trait;
+use  Google\Site_Kit\Modules\Analytics\AccountTicket;
+/**
+ * Class for Analytics provisioning.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Provisioning extends \Google\Site_Kit_Dependencies\Google_Service_Resource {
```
Comment (felixarntz) [inline includes/Modules/Analytics/Provisioning.php:23]: Same here, maybe `Proxy_Provisioning`? Also better add a `use` statement for that service class rather than having the full name here.

File: includes/Modules/Analytics/Provisioning.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	/**
+	 * Creates an account ticket. (provisioning.createAccountTicket)
+	 *
+	 * @param Google_Service_Analytics_AccountTicket $post_body The post body to send.
```
Comment (felixarntz) [inline includes/Modules/Analytics/Provisioning.php:27]: Minor indentation issue here. Parameter descriptions should be intented the same way. Also the type here is not correct, it needs to be _our_ `AccountTicket` class.

File: includes/Modules/Analytics/Provisioning.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+ * @copyright 2019 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Modules\Analytics;
+
+use Google\Site_Kit\Core\Modules\Module_Settings;
+use Google\Site_Kit\Core\Storage\Setting_With_Legacy_Keys_Trait;
+use  Google\Site_Kit\Modules\Analytics\AccountTicket;
```
Comment (felixarntz) [inline includes/Modules/Analytics/Provisioning.php:15]: One space too much.

File: includes/Modules/Analytics/Provisioning.php [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+/**
```
Comment (felixarntz) [inline includes/Modules/Analytics/Provisioning.php:16]: There should be a blank line above this.

File: includes/Modules/Analytics/Provisioning.php [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+	 * Creates an account ticket. (provisioning.createAccountTicket)
+	 *
+	 * @param Google_Service_Analytics_AccountTicket $post_body The post body to send.
+	 * @param array                                  $opt_params Optional parameters. Pass site_id and site_secret here.
+	 * @return Google_Service_Analytics_AccountTicket
+	 */
+	public function createAccountTicket( AccountTicket $post_body, $opt_params = array() ) {
+		$params = array( 'postBody' => $post_body );
+		$params = \array_merge( $params, $opt_params );
+		return $this->call( 'createAccountTicket', array( $params ), 'Google\Site_Kit\Modules\Analytics\AccountTicket' );
```
Comment (felixarntz) [inline includes/Modules/Analytics/Provisioning.php:34]: I believe here you shouldn't need to pass _our_ `AccountTicket` class name. It's for the response type, and that will still be a regular `Google_Service_Analytics_AccountTicket`. We only need the custom `AccountTicket` for our request.

File: includes/Modules/Analytics.php [hunk 13]
```diff
@@ -767,6 +775,86 @@ function ( $order_def ) {
...
+									'path'       => 'provisioning/createAccountTicket',
+									'httpMethod' => 'POST',
+									'parameters' => array(),
+								),
+							),
+						)
+					);
+
+					// Use a custom provisioning method that accepts site id and secret.
+					$analytics_service->provisioning = $custom_provisioning;
```
[the remark this answers is not in the document]
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:842]: 👍 great idea, moved.

File: includes/Modules/Analytics.php (hunk 2 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:797]: fixed

File: includes/Modules/Analytics.php (hunk 3 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:795]: Doh! I though I tried that. Changed.

File: includes/Modules/Analytics.php (hunk 4 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:491]: fixed.

File: includes/Modules/Analytics.php (hunk 5 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:853]: removed

File: includes/Modules/Analytics.php (hunk 6 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:857]: Ah right, thanks for clarifying... changing this.

File: includes/Modules/Analytics/Account_Ticket.php (hunk 7 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics/Account_Ticket.php:22]: yea, was wondering about that as well, good suggestion. renamed

File: includes/Modules/Analytics/Provisioning.php (hunk 8 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics/Provisioning.php:23]: changed, moved to use

File: includes/Modules/Analytics/Provisioning.php (hunk 10 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics/Provisioning.php:15]: fixed

File: includes/Modules/Analytics/Provisioning.php (hunk 11 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics/Provisioning.php:16]: added

File: includes/Modules/Analytics/Provisioning.php (hunk 12 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics/Provisioning.php:34]: good catch, changed.

**Comment by adamsilverstein** [comment]
@felixarntz thanks for the tips and feedback, ready for another review.

File: includes/Modules/Analytics.php [hunk 14]
```diff
@@ -1236,6 +1326,17 @@ protected function setup_settings() {
...
 
+	/**
+	 * Gets the provisioning redirect URI that listens for the Terms of Service redirect.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return string Provisioning redirect URI.
+	 */
+	private function get_provisioning_redirect_uri() {
+		return add_query_arg( 'gatoscallback', '1', untrailingslashit( home_url() ) );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:1337]: @felixarntz should we be introducing new boolean query params to listen for or does the API require this param specifically? I feel like we could get by just as well with a single parameter such as `googlesitekit-callback={callback-id}`. This would likely cover most of our cases, since I don't think we ever need to run multiple callbacks of this type in the same request and avoid potential conflicts with other plugins in the future.

Comment (felixarntz) [reply includes/Modules/Analytics.php:1337]: I agree with what you're saying, but we've already defined this in several places, on the authentication service too. It's also in line with our other callback URL, which isn't really a point, but I think it's okay to use the parameter like this.

It's not a big deal to me, the only reason I'd say we shouldn't change it despite your argument is because it would require us to make changes across several places.

File: includes/Modules/Analytics.php [hunk 15]
```diff
@@ -1128,9 +1195,32 @@ private function get_analyticsreporting_service() {
...
+				'methods' => array(
+					'createAccountTicket' => array(
+						'path'       => 'provisioning/createAccountTicket',
+						'httpMethod' => 'POST',
+						'parameters' => array(),
+					),
+				),
+			)
+		);
+		$analytics_provisioning_service->provisioning = $custom_provisioning;
```
[the remark this answers is not in the document]
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:1218]: good point, updated.

File: includes/Modules/Analytics.php (hunk 6 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:828]: Ah, right. Removed this bit and calling/returning at once.

File: includes/Modules/Analytics/Proxy_Account_Ticket.php [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+	/**
+	 * The site secret.
+	 *
+	 * @since n.e.x.t
+	 * @var String
+	 */
+	public $site_secret = '';
+
+	/**
+	 * Get the site ID.
```
[the remark this answers is not in the document]
Comment (adamsilverstein) [reply includes/Modules/Analytics/Proxy_Account_Ticket.php]: fixed

File: includes/Modules/Analytics/Proxy_Account_Ticket.php [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+	 *
+	 * @since n.e.x.t
+	 */
+	public function getSiteId() {
+		return $this->site_id;
+	}
+
+	/**
+	 * Set the site ID.
```
[the remark this answers is not in the document]
Comment (adamsilverstein) [reply includes/Modules/Analytics/Proxy_Account_Ticket.php]: updated

**Comment by adamsilverstein** [comment]
Addressed feedback and cleared up merge conflicts with `feature/ga-provisioning`.

File: includes/Modules/Analytics/Proxy_Provisioning.php [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+	/**
+	 * Creates an account ticket. (provisioning.createAccountTicket)
+	 *
+	 * @param Proxy_AccountTicket $post_body  The post body to send.
+	 * @param array               $opt_params Optional parameters. Pass site_id and site_secret here.
+	 * @return Google_Service_Analytics_AccountTicket
+	 */
+	public function createAccountTicket( Proxy_AccountTicket $post_body, $opt_params = array() ) {
+		$params = array( 'postBody' => $post_body );
+		$params = \array_merge( $params, $opt_params );
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics/Proxy_Provisioning.php:35]: Let's avoid prefixing functions with `\`. If really needed, we can `use` functions rather than using fully-qualified paths inline.

File: includes/Modules/Analytics/Proxy_Account_Ticket.php [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+	 */
+	public $site_id = '';
+
+	/**
+	 * The site secret.
+	 *
+	 * @since n.e.x.t
+	 * @var String
+	 */
+	public $site_secret = '';
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics/Proxy_Account_Ticket.php:38]: @adamsilverstein why are these public? It seems odd when there are getters and setters for them. Is this related to how the underlying classes work?

File: includes/Modules/Analytics.php [hunk 20]
```diff
@@ -51,6 +54,8 @@
 use WP_Error;
 use Exception;
 
+const ANALYTICS_PROVISIONING_ACCOUNT_TICKET_ID = 'googlesitekit_analytics_provision_atid';
```
Comment (aaemnnosttv) [inline includes/Modules/Analytics.php:57]: As a class constant, this could should drop the `ANALYTICS_` prefix.

I also think we should avoid aggressive abbreviation with the key, unless there is a hard limit somewhere.

How about:

```php
const PROVISION_ACCOUNT_TICKET_ID = 'googlesitekit_analytics_provision_account_ticket_id'
```
It's a little longer, but I think the readability is worth it.

File: includes/Modules/Analytics/Proxy_Provisioning.php (hunk 18 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics/Proxy_Provisioning.php:35]: 👍 will change

File: includes/Modules/Analytics/Proxy_Account_Ticket.php (hunk 19 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics/Proxy_Account_Ticket.php:38]: yes, this is because the public elements are sent when the account ticket is sent in the api request. The setters and getters follow the pattern from default `Account_Ticket` class.

File: includes/Modules/Analytics.php (hunk 20 above)
Comment (adamsilverstein) [reply includes/Modules/Analytics.php:57]: 👍
